src/find_sort_location.rs

```diff
@@ -0,0 +1,142 @@
+#![allow(dead_code)]
+
+use futures::future::join_all;
+
+use crate::{
+    field::Field,
+    helpers::ring::Ring,
+    replicated_secret_sharing::ReplicatedSecretSharing,
+    securemul::{ProtocolContext, SecureMul},
+};
+
+pub struct SecureSort<'a, R, F> {
+    context: &'a ProtocolContext<'a, R>,
+    share_of_one: ReplicatedSecretSharing<F>,
+}
+
+impl<F: Field, R: Ring> SecureSort<'_, R, F> {
+    pub async fn execute(
+        self,
+        shares: &[ReplicatedSecretSharing<F>],
+    ) -> Vec<ReplicatedSecretSharing<F>> {
+        let gen_permutation_share = |mult_output: Vec<ReplicatedSecretSharing<F>>| {
```

Comment:
do you think it is worth adding `debug_assert` to make sure `mult_output` length is even? seems that `gen_permutation_share` will be broken if that invariant does not hold